### PR TITLE
Particle effect when quarry digs a node

### PR DESCRIPTION
**Pull request by JordanL2**
I thought it would be nice if there was more visual feedback when the quarry digs a node.

Now each time a node is dug, we get a small explosion of magenta particles.

### CODE REVIEW

**Comment by OgelGames** [comment]
👎 This is bad for performance, both for the server and any players nearby. Also, quarries are not usually watched much, so it would be a waste most of the time.

It might be useful though for the current node to be shown along with the other visuals when a quarry is punched.

**Comment by JordanL2** [comment]
The quarry digs one block a second. What kind of performance impact would be acceptable? The particles last 0.5 seconds at most.

**Comment by OgelGames** [comment]
It's not just the particles that have an impact on performance, but also the processing time it takes to create them. The more "stuff" a quarry has to do each time it runs, the less quarries can be used on a network before lag-reduction is triggered.

I'd like to hear @S-S-X's opinion on this, he knows more about particles than me.

**Comment by S-S-X** [comment]
This would be very nice feature but unfortunately particles have to be sent to every user and hundreds quarries digging nodes every second causes a lot of network traffic even while other server load would not be that significant:

Assuming powerful server and Technic not rate limiting networks itself because of lag.
Let's assume that there's 200 active quarries and 15 players connected. (based on: I've build and used digger setups with hundred quarries each and know that some players built a lot larger).
Also last spike for potentially active quarries for pandorabox during last 24 hours was 306 quarries.

What happens every second:
* 200 nodes dug.
* 15 * 200 particle effects sent to connected players.
* Extra outgoing network packets per second: 3000 pps

AFAIK engine does not optimize particle effects currently and therefore this would need some optimization to ignore players who are not nearby, doing that increases server side workload on main thread but reduces network load (and of course network thread load).

In practice optimizing effects only for nearby (quarry dig position) players could make it work but straight loop over players to check positions wont be good idea, it has to use better techniques and preferably build cache groups for nearby quarries.
One simple way could be to just use mapchunks or some other large grid like that but of course any algorithm that can keep engine load low would be fine.

### Note:
I told you _"AFAIK engine does not optimize particle effects currently"_, you could verify that and come back with evidence if particles are actually optimized well enough to allow huge amounts to be spawned at any given time.

**Comment by JordanL2** [comment]
@S-S-X take a look at this merged PR from 2017:
https://github.com/minetest/minetest/pull/5760
Seems to be saying that particles are only sent to players if they're close enough to see them.

**Comment by S-S-X** [comment]
> @S-S-X take a look at this merged PR from 2017:
> [minetest/minetest#5760](https://github.com/minetest/minetest/pull/5760)
> Seems to be saying that particles are only sent to players if they're close enough to see them.

Thanks for looking up, seems like builtin optimizations is there. That does change situation a lot as most of time quarries are digging at positions further away from players and therefore only small amount of particles will be actually sent.

It might still cause problems in some situations but it is a lot less likely and should not affect all players.
For large quarry arrays it can still increase outgoing traffic by few hundred packets per second which is bad but knowing that engine has some optimizations built in it will be a lot easier to limit that.

It would still be good to limit particles a bit, for example just random chance could work well enough in most situations and is extremely simple to implement. There could also be network time usage check that allows spawning particles only if average time usage is at most 75% of throttling line for example.
Also there should be server wide configuration option for this to enable/disable effects.

TLDR what I think is needed:
* Add configuration, default disabled is easier to get merged (technic.conf)
* See if there's easy / simple enough way to rate limit particles at least when controlling network reports high time usage.
    * Possibly dynamic rate limits.
    * Possibly also static mapblock / node neighbor / distance / etc based rate limits to not allow 1000 quarries at same location spawn 1000 particles at once.

@OgelGames what do you think about it if above TLDR list would be implemented?
Particles are cheap for server itself, those are basically fire and forget server side (there's no significant processing involved at all).
For clients if server is sending too many particles it can affect players behind slower internet connections. Rate limits should solve this.

In my opinion unnecessary effects like this are good addition to any game but there should always be a way to disable those and some care should be taken to keep it lightweight / limit resource usage and to not affect things that provide actual necessary functionality.

**Comment by S-S-X** [comment]
Also @JordanL2 could you come up with short lived particle spawners (or even few if need) instead of sending 50 individual particles.

**Comment by JordanL2** [comment]
> Also @JordanL2 could you come up with short lived particle spawners (or even few if need) instead of sending 50 individual particles? Those can randomize stuff client side and kind of send multiple particles at once reducing server side workload a lot (and reduces relatively slow "system calls" to C++ side).

The problem is that I ensure the particles all have a velocity of 3, by randomly generating the x,y,z velocity values and then normalising them so the magnitude of the vector is 3. I can't see a way of doing that with the particle spawner.

I've just pushed a commit lowering the particle count to 20 (still looks good to me) and adding the option to enable the particles (`enable_quarry_dig_particles`), disabled by default.

**Comment by JordanL2** [comment]
I just came up with what I think is an even better looking effect with just 8 particles. The particles start in the 8 corners of the cube, and move towards the middle, disappearing when they reach the exact middle. Looks a bit like the node has been teleported away or something.

> at least when controlling network reports high time usage

You mean the electricity network the quarry is attached to? I'll look into that.

**Comment by S-S-X** [comment]
> You mean the electricity network the quarry is attached to? I'll look into that.

Full signature and average wasted time example for `technic_run` callback used when registering technic mahcine nodes:

```lua
nodedef.technic_run = function(pos, node, run_stage, network)
  if network.lag < 35000 then
    print("Lag seems fine for this network")
  end
end
```

**Comment by JordanL2** [comment]
Commit pushed so particles are only generated if lag < 35000.

**Comment by JordanL2** [comment]
Part of me is wondering if users should just disable particles if they've got 1000 quarries in one place, or getting significant lag.

**Comment by OgelGames** [comment]
> @OgelGames what do you think about it if above TLDR list would be implemented?

I think it would be fine. I definitely agree with adding a setting for it, but I'm not sure if dynamic rate-limiting is a good thing to add.

Also, how about making the particles the same as the ones created when a player digs a node? It would not only look nicer, but can also be done with a particle spawner:
```lua
minetest.add_particlespawner({
	amount = 16,
	time = 0.01,
	minpos = vector.subtract(dig_pos, 0.5),
	maxpos = vector.add(dig_pos, 0.5),
	minvel = {x =-0.5, y =-1, z =-0.5},
	maxvel = {x = 0.5, y = 0, z = 0.5},
	minacc = {x = 0, y = -9.81, z = 0},
	maxacc = {x = 0, y = -9.81, z = 0},
	minsize = 0,
	maxsize = 0,
	node = dig_node
})
```

**Comment by JordanL2** [comment]
@OgelGames that gave me an idea for an even better effect - the node's particles now stream towards the quarry itself (stopping just as they reach it), like the quarry is vacuuming up the node. This uses a particle spawner.

It has 50 particles over 0.5 seconds, I'm not sure how much particle counts lag clients. If the player is standing next to 100 quarries then 5000 particles could be a bit much, but since particles are only sent to players if they're "close enough", I'm not sure how many quarries a player can possibly be standing close enough to at once. It seems it's determined by `max_block_send_distance` which seems to default to 10, so I suppose it is possible to be within range of 100 of them.

**Comment by OgelGames** [comment]
> the node's particles now stream towards the quarry itself (stopping just as they reach it), like the quarry is vacuuming up the node

I like that, adding a tiny bit of logic to the magical quarry :)

**Comment by JordanL2** [comment]
> but I'm not sure if dynamic rate-limiting is a good thing to add

Currently, particles aren't sent if the network lag is more than 35000 (number taken from example above). I'm not sure if any more limiting should be applied.

File: technic/machines/HV/quarry.lua
Comment (OgelGames) [inline technic/machines/HV/quarry.lua:247]: These calculations could probably be done simpler (and faster) using the `vector.*` functions: [lua_api.txt#L3325-L3376](https://github.com/minetest/minetest/blob/97248c695730c29ca91fb7e56456d061f77744f5/doc/lua_api.txt#L3325-L3376)

Comment (JordanL2) [reply technic/machines/HV/quarry.lua]: Commit pushed.

**Review by OgelGames — APPROVED** [review]
Tested, works and looks good, and doesn't impact quarry performance much 👍

**Comment by JordanL2** [comment]
I'll quickly squash the commits in one.

**Comment by OgelGames** [comment]
> I'll quickly squash the commits in one.

The PR will be squashed anyway :)

**Comment by S-S-X** [comment]
Note that `should_generate_particles` could still be dropped and `net` variable forwarded through from quarry_run where it wont need any validation.
That's just cleaning up, I think around -6 lines, -1 function and -3 variable checks.

**Comment by JordanL2** [comment]
> Note that `should_generate_particles` could still be dropped and `net` variable forwarded through from `quarry_run` where it wont need any validation. That's just cleaning up, I think around -9 lines, -1 function and -3 variable checks.

I tried expanding the signature for quarry_run like so:
```
local function quarry_run(pos, node, run_stage, network)
```

network is nil though. Which is weird because it's exactly the same as the example in battery_box.lua:
```
local function run(pos, node, run_state, network)
```

**Comment by S-S-X** [comment]
> network is nil though. Which is weird because it's exactly the same as the example in battery_box.lua:

Well, sorry about that. Actually my fault as I've only added that for battery boxes and did not remember doing it that way...
Was not sure about API back then and decided to only expose it for battery boxes where it was immediately used for battery box optimizations.

I think it could be added for everything else too. I think API for `network` instances should be stable enough to do that if you want. Basically just add `network` argument here:
https://github.com/mt-mods/technic/blob/7e466e9b41215306b220d83b932b754294f32e1d/technic/machines/network.lua#L623-L625

That's only place where modification is needed, basically just add `network` for 2 run_nodes similar how it is for BA_nodes.

**Comment by JordanL2** [comment]
@S-S-X will do, to be clear you want me to add the network argument to these two calls?
```
 run_nodes(PR_nodes, vm, technic.producer) 
 run_nodes(RE_nodes, vm, technic.receiver) 
```

On another node, I've just committed a change to how the particles approach the quarry. They now accelerate along a particular axis (depending on the quarry's orientation, e.g. if it's facing north they accelerate on the Z axis). The effect is that they approach in a curve, which looks much nicer and means when digging nodes that are far down the particles won't be travelling through the cliff face. The acceleration also enhances the impression the quarry is vacuuming up the particles.

**Comment by JordanL2** [comment]
Commit pushed for network arg.

**Comment by JordanL2** [comment]
Also, regarding the new vector code - For some reason the functions `vector.copy` and `vector.zero` don't seem to be available despite being listed in lua_api.txt, am I missing something? Anyway that's why I copied a vector and created a zero vector the dumb way.

**Comment by S-S-X** [comment]
> For some reason the functions `vector.copy` and `vector.zero` don't seem to be available despite being listed in lua_api.txt, am I missing something?

Remember to use either minimum supported version API: https://raw.githubusercontent.com/minetest/minetest/5.0.0/doc/lua_api.txt
or at least latest **official release** version API:
https://raw.githubusercontent.com/minetest/minetest/5.4.1/doc/lua_api.txt

**Comment by JordanL2** [comment]
Thanks that's helpful, I've pushed a commit cleaning up the code a bit.

What do you think of the effect with the curve / acceleration?

**Comment by OgelGames** [comment]
> What do you think of the effect with the curve / acceleration?

I think it does look better 👍

File: technic/machines/HV/quarry.lua
Comment (OgelGames) [inline technic/machines/HV/quarry.lua:236]:
```suggestion
	quarry_pos = vector.add(quarry_pos, minetest.facedir_to_dir(param2))
```

**Comment by JordanL2** [comment]
A couple of questions:
1. Do we need to update docs so people know about the `enable_quarry_dig_particles` option?
2. Currently this option defaults to false. Should it default to true?

**Comment by OgelGames** [comment]
> 1. Do we need to update docs so people know about the `enable_quarry_dig_particles` option?

That would be good to do, yes. Should only need an extra line in this section of the readme: https://github.com/mt-mods/technic#settings-worldpathtechnicconf

> 2. Currently this option defaults to false. Should it default to true?

I don't think so, it's better to have it opt-it IMO.

**Comment by JordanL2** [comment]
Ok, added a line to README.md.

**Review by S-S-X — APPROVED** [review]
All Mineunit tests passed fine, it is again just inability to post test coverage report because PR is coming from forked repository.
Did not test but LGTM.

**Comment by S-S-X** [comment]
Fine for me either way but one thing that could be still changed: configuration key name.
```lua
	quarry_dig_above_nodes = "3",
	quarry_max_depth = "100",
	quarry_time_limit = "5000",
	enable_quarry_dig_particles = "false",
```
to something like:
```lua
	quarry_dig_above_nodes = "3",
	quarry_max_depth = "100",
	quarry_time_limit = "5000",
	quarry_dig_particles = "false",
```

**Comment by JordanL2** [comment]
Will make that change now.

**Comment by OgelGames** [comment]
> @OgelGames anything left to do for this issue, any further concerns or changes needed?

No, all good by me 👍

**Comment by JordanL2** [comment]
Renamed enable_quarry_dig_particles to quarry_dig_particles, tested it still works.

**Comment by S-S-X** [comment]
`    technic/config.lua:30:2: expected '}' (to close '{' on line 3) near 'switch_off_delay_seconds'`

**Comment by S-S-X** [comment]
Btw, finally added possible fix for Mineunit coverage report failure: https://github.com/mt-mods/technic/pull/258
Can be either merged before this PR and tested with this PR or can be done later.

**Comment by S-S-X** [comment]
@JordanL2 do you want to rebase yourself? Not needed but if you want you can do so.
Mineunit coverage report thing fixed on master so all workflows are now passing.

**Comment by JordanL2** [comment]
I think squashing into a single commit would be fine.

**Comment by S-S-X** [comment]
Will be merging this very soon unless someone complains.

**Comment by S-S-X** [comment]
Got third approval and recommendation to merge right away. Thanks @JordanL2

**Comment by BuckarooBanzay** [comment]
too late for the party, sorry. Looks nice though :+1:

**Pull request merged**
